File: ios/Sources/RommioFoundation/PlatformControlProfileResolver.swift
import Foundation

public struct PlatformControlProfileResolver: ControlProfileResolving {
    private let profiles: [PlatformControlProfile]

    public init(profiles: [PlatformControlProfile]? = nil) {
        self.profiles = profiles ?? Self.defaultProfiles
    }

    public func resolve(platformSlug: String) -> PlatformControlProfile {
        let normalized = platformSlug.lowercased()
        return profiles.first(where: { $0.platformSlugs.contains(normalized) }) ?? unsupportedProfile(
            familyID: normalized,
            displayName: normalized.replacingOccurrences(of: "-", with: " ").capitalized,
            platformSlugs: [normalized],
            message: "This platform is not enabled in the embedded player yet."
        )
    }

    public func supportedProfiles() -> [PlatformControlProfile] {
        profiles
    }

    public static let defaultProfiles: [PlatformControlProfile] = [
        touchProfile(
            familyID: "nes",
            displayName: "NES / Famicom",
            platformSlugs: ["nes", "famicom"],
            presets: [
                TouchLayoutPreset(
                    presetID: "standard",
                    displayName: "Standard",
                    elements: [
                        standardDpad(),
                        faceTwo(
                            primaryLabel: "A",
                            secondaryLabel: "B",
                            primaryKey: LibretroJoypadID.a,
                            secondaryKey: LibretroJoypadID.b
                        ),
                        standardStartSelect(),
                    ]
                ),
                TouchLayoutPreset(
                    presetID: "compact",
                    displayName: "Compact",
                    elements: [
                        standardDpad().copy(centerX: 0.16, centerY: 0.76, baseScale: 0.9),
                        faceTwo(
                            primaryLabel: "A",
                            secondaryLabel: "B",
                            primaryKey: LibretroJoypadID.a,
                            secondaryKey: LibretroJoypadID.b
                        ).copy(centerX: 0.84, centerY: 0.75, baseScale: 0.92),
                        standardStartSelect().copy(centerY: 0.86, baseScale: 0.82),
                    ]
                ),
            ]
        ),
        touchProfile(
            familyID: "snes",
            displayName: "SNES / Super Famicom",
            platformSlugs: ["snes", "sfam"],
            orientationPolicy: .landscapeOnly,
            presets: [
                TouchLayoutPreset(
                    presetID: "standard",
                    displayName: "Standard",
                    elements: [
                        standardDpad(),
                        faceFour(
                            left: ("Y", LibretroJoypadID.y),
                            bottom: ("B", LibretroJoypadID.b),
                            right: ("A", LibretroJoypadID.a),
                            top: ("X", LibretroJoypadID.x)
                        ),
                        standardShoulders(labels: ("L", "R")),
                        standardStartSelect(),
                    ]
                ),
            ]
        ),
        touchProfile(
            familyID: "gb",
            displayName: "Game Boy / Color",
            platformSlugs: ["gb", "gbc"],
            presets: [
                TouchLayoutPreset(
                    presetID: "standard",
                    displayName: "Standard",
                    elements: [
                        standardDpad(),
                        faceTwo(
                            primaryLabel: "A",
                            secondaryLabel: "B",
                            primaryKey: LibretroJoypadID.a,
                            secondaryKey: LibretroJoypadID.b
                        ),
                        standardStartSelect(),
                    ]
                ),
            ]
        ),
        touchProfile(
            familyID: "gba",
            displayName: "Game Boy Advance",
            platformSlugs: ["gba"],
            orientationPolicy: .landscapeOnly,
            presets: [
                TouchLayoutPreset(
                    presetID: "standard",
                    displayName: "Standard",
                    elements: [
                        standardDpad(),
                        faceTwo(
                            primaryLabel: "A",
                            secondaryLabel: "B",
                            primaryKey: LibretroJoypadID.a,
                            secondaryKey: LibretroJoypadID.b
                        ).copy(centerY: 0.70),
                        standardShoulders(labels: ("L", "R")),
                        standardStartSelect(),
                    ]
                ),
            ]
        ),
        touchProfile(
            familyID: "sega16",
            displayName: "Sega 8/16-bit",
            platformSlugs: ["genesis", "sms", "gamegear", "segacd"],
            presets: [
                TouchLayoutPreset(
                    presetID: "standard",
                    displayName: "Standard",
                    elements: [
                        standardDpad(),
                        faceFour(
                            left: ("B", LibretroJoypadID.y),
                            bottom: ("A", LibretroJoypadID.b),
                            right: ("C", LibretroJoypadID.a),
                            top: ("X", LibretroJoypadID.x)
                        ).copy(centerY: 0.69),
                        standardStartSelect(selectLabel: "Mode", startLabel: "Start"),
                    ]
                ),
            ]
        ),
        controllerFirstProfile(
            familyID: "sega32x",
            displayName: "Sega 32X",
            platformSlugs: ["32x", "sega32x", "sega-32x"]
        ),
        touchProfile(
            familyID: "psx",
            displayName: "PlayStation",
            platformSlugs: ["psx", "ps1", "playstation"],
            orientationPolicy: .landscapeOnly,
            presets: [
                TouchLayoutPreset(
                    presetID: "digital",
                    displayName: "Digital",
                    elements: [
                        standardDpad(),
                        faceFour(
                            left: ("Square", LibretroJoypadID.y),
                            bottom: ("Cross", LibretroJoypadID.b),
                            right: ("Circle", LibretroJoypadID.a),
                            top: ("Triangle", LibretroJoypadID.x)
                        ),
                        standardShoulders(
                            id: "shoulders_top",
                            labels: ("L1", "R1"),
                            keyCodes: (LibretroJoypadID.l1, LibretroJoypadID.r1)
                        ),
                        standardShoulders(
                            id: "shoulders_bottom",
                            labels: ("L2", "R2"),
                            keyCodes: (LibretroJoypadID.l2, LibretroJoypadID.r2)
                        ).copy(centerY: 0.25, baseScale: 0.84),
                        standardStartSelect(),
                    ]
                ),
            ]
        ),
        touchProfile(
            familyID: "atari",
            displayName: "Atari / Handheld Classics",
            platformSlugs: ["atari2600", "atari7800", "lynx"],
            presets: [
                TouchLayoutPreset(
                    presetID: "standard",
                    displayName: "Standard",
                    elements: [
                        standardDpad(),
                        faceTwo(
                            primaryLabel: "1",
                            secondaryLabel: "2",
                            primaryKey: LibretroJoypadID.a,
                            secondaryKey: LibretroJoypadID.b
                        ),
                        standardStartSelect(selectLabel: "Option", startLabel: "Start"),
                    ]
                ),
            ]
        ),
        touchProfile(
            familyID: "tg16",
            displayName: "TurboGrafx / Neo Geo Pocket / WonderSwan",
            platformSlugs: ["tg16", "neo-geo-pocket", "neo-geo-pocket-color", "wonderswan", "wonderswan-color"],
            presets: [
                TouchLayoutPreset(
                    presetID: "standard",
                    displayName: "Standard",
                    elements: [
                        standardDpad(),
                        faceTwo(
                            primaryLabel: "II",
                            secondaryLabel: "I",
                            primaryKey: LibretroJoypadID.a,
                            secondaryKey: LibretroJoypadID.b
                        ),
                        standardStartSelect(),
                    ]
                ),
            ]
        ),
        touchProfile(
            familyID: "arcade",
            displayName: "Arcade",
            platformSlugs: ["arcade"],
            presets: [
                TouchLayoutPreset(
                    presetID: "standard",
                    displayName: "Standard",
                    elements: [
                        standardDpad(),
                        faceFour(
                            left: ("1", LibretroJoypadID.y),
                            bottom: ("2", LibretroJoypadID.b),
                            right: ("3", LibretroJoypadID.a),
                            top: ("4", LibretroJoypadID.x)
                        ),
                        standardStartSelect(selectLabel: "Coin", startLabel: "Start"),
                    ]
                ),
            ]
        ),
        controllerFirstProfile(
            familyID: "n64",
            displayName: "Nintendo 64",
            platformSlugs: ["n64"]
        ),
        controllerFirstProfile(
            familyID: "psp",
            displayName: "PlayStation Portable",
            platformSlugs: ["psp"]
        ),
        touchProfile(
            familyID: "nds",
            displayName: "Nintendo DS / DSi-enhanced",
            platformSlugs: ["nds", "dsi", "nintendo-dsi"],
            orientationPolicy: .portraitOnly,
            presets: [
                TouchLayoutPreset(
                    presetID: "portrait-handheld",
                    displayName: "Portrait handheld",
                    elements: [
                        standardDpad().copy(centerX: 0.19, centerY: 0.80),
                        faceFour(
                            left: ("Y", LibretroJoypadID.y),
                            bottom: ("B", LibretroJoypadID.b),
                            right: ("A", LibretroJoypadID.a),
                            top: ("X", LibretroJoypadID.x)
                        ).copy(centerX: 0.81, centerY: 0.79, baseScale: 0.94),
                        standardShoulders(labels: ("L", "R")).copy(centerY: 0.11, baseScale: 0.86),
                        standardStartSelect().copy(centerY: 0.72, baseScale: 0.82),
                    ]
                ),
            ]
        ),
        controllerFirstProfile(
            familyID: "dreamcast",
            displayName: "Dreamcast / NAOMI",
            platformSlugs: ["dreamcast", "naomi"]
        ),
        controllerFirstProfile(
            familyID: "3do",
            displayName: "3DO",
            platformSlugs: ["3do"]
        ),
        controllerFirstProfile(
            familyID: "virtualboy",
            displayName: "Virtual Boy",
            platformSlugs: ["virtualboy", "virtual-boy"]
        ),
        controllerFirstProfile(
            familyID: "dolphin",
            displayName: "GameCube / Wii",
            platformSlugs: ["gamecube", "ngc", "wii"]
        ),
        controllerFirstProfile(
            familyID: "3ds",
            displayName: "Nintendo 3DS",
            platformSlugs: ["3ds", "new-nintendo-3ds"]
        ),
        controllerFirstProfile(
            familyID: "dos",
            displayName: "DOS",
            platformSlugs: ["dos"],
            message: "DOS stays controller-first until a validated pointer and keyboard helper overlay exists."
        ),
    ]
}

private func touchProfile(
    familyID: String,
    displayName: String,
    platformSlugs: Set<String>,
    orientationPolicy: PlayerOrientationPolicy = .auto,
    presets: [TouchLayoutPreset]
) -> PlatformControlProfile {
    let preferredViewportAspectRatio: Double?
    switch familyID {
    case "nes", "snes", "sega16", "psx", "atari", "tg16", "arcade":
        preferredViewportAspectRatio = 4.0 / 3.0
    case "gb":
        preferredViewportAspectRatio = 10.0 / 9.0
    case "gba":
        preferredViewportAspectRatio = 3.0 / 2.0
    case "nds":
        preferredViewportAspectRatio = 2.0 / 3.0
    default:
        preferredViewportAspectRatio = nil
    }

    return PlatformControlProfile(
        familyID: familyID,
        displayName: displayName,
        platformSlugs: platformSlugs,
        supportTier: .touchSupported,
        touchSupportMode: .full,
        playerOrientationPolicy: orientationPolicy,
        preferredViewportAspectRatio: preferredViewportAspectRatio,
        defaultPresetID: presets.first?.presetID,
        presets: presets
    )
}

private func controllerFirstProfile(
    familyID: String,
    displayName: String,
    platformSlugs: Set<String>,
    message: String = "Connect a controller for this platform. Touch support comes later after a platform-specific layout is validated."
) -> PlatformControlProfile {
    PlatformControlProfile(
        familyID: familyID,
        displayName: displayName,
        platformSlugs: platformSlugs,
        supportTier: .controllerSupported,
        touchSupportMode: .controllerFirst,
        controllerFallbackMessage: message
    )
}

private func unsupportedProfile(
    familyID: String,
    displayName: String,
    platformSlugs: Set<String>,
    message: String
) -> PlatformControlProfile {
    PlatformControlProfile(
        familyID: familyID,
        displayName: displayName,
        platformSlugs: platformSlugs,
        supportTier: .unsupported,
        touchSupportMode: .controllerFirst,
        controllerFallbackMessage: message
    )
}

private extension TouchElementSpec {
    func copy(
        centerX: Double? = nil,
        centerY: Double? = nil,
        baseScale: Double? = nil
    ) -> TouchElementSpec {
        TouchElementSpec(
            id: id,
            label: label,
            layoutKind: layoutKind,
            buttons: buttons,
            centerX: centerX ?? self.centerX,
            centerY: centerY ?? self.centerY,
            baseScale: baseScale ?? self.baseScale
        )
    }
}

File: app/src/main/java/io/github/bandoracer/rommio/domain/input/PlatformControlProfileResolver.kt
package io.github.bandoracer.rommio.domain.input

import android.view.KeyEvent
import io.github.bandoracer.rommio.domain.player.EmbeddedSupportTier

class PlatformControlProfileResolver : ControlProfileResolver {
    private val profiles = listOf(
        touchProfile(
            familyId = "nes",
            displayName = "NES / Famicom",
            platformSlugs = setOf("nes", "famicom"),
            presets = listOf(
                TouchLayoutPreset(
                    presetId = "standard",
                    displayName = "Standard",
                    elements = listOf(
                        standardDpad(),
                        faceTwo(
                            primaryLabel = "A",
                            secondaryLabel = "B",
                            primaryKey = KeyEvent.KEYCODE_BUTTON_A,
                            secondaryKey = KeyEvent.KEYCODE_BUTTON_B,
                        ),
                        standardStartSelect(),
                    ),
                ),
                TouchLayoutPreset(
                    presetId = "compact",
                    displayName = "Compact",
                    elements = listOf(
                        standardDpad().copy(centerX = 0.16f, centerY = 0.76f, baseScale = 0.9f),
                        faceTwo(
                            primaryLabel = "A",
                            secondaryLabel = "B",
                            primaryKey = KeyEvent.KEYCODE_BUTTON_A,
                            secondaryKey = KeyEvent.KEYCODE_BUTTON_B,
                        ).copy(centerX = 0.84f, centerY = 0.75f, baseScale = 0.92f),
                        standardStartSelect().copy(centerY = 0.86f, baseScale = 0.82f),
                    ),
                ),
            ),
        ),
        touchProfile(
            familyId = "snes",
            displayName = "SNES / Super Famicom",
            platformSlugs = setOf("snes", "sfam"),
            orientationPolicy = PlayerOrientationPolicy.LANDSCAPE_ONLY,
            presets = listOf(
                TouchLayoutPreset(
                    presetId = "standard",
                    displayName = "Standard",
                    elements = listOf(
                        standardDpad(),
                        faceFour(
                            left = "Y" to KeyEvent.KEYCODE_BUTTON_Y,
                            bottom = "B" to KeyEvent.KEYCODE_BUTTON_B,
                            right = "A" to KeyEvent.KEYCODE_BUTTON_A,
                            top = "X" to KeyEvent.KEYCODE_BUTTON_X,
                        ),
                        standardShoulders(labels = "L" to "R"),
                        standardStartSelect(),
                    ),
                ),
            ),
        ),
        touchProfile(
            familyId = "gb",
            displayName = "Game Boy / Color",
            platformSlugs = setOf("gb", "gbc"),
            presets = listOf(
                TouchLayoutPreset(
                    presetId = "standard",
                    displayName = "Standard",
                    elements = listOf(
                        standardDpad(),
                        faceTwo(
                            primaryLabel = "A",
                            secondaryLabel = "B",
                            primaryKey = KeyEvent.KEYCODE_BUTTON_A,
                            secondaryKey = KeyEvent.KEYCODE_BUTTON_B,
                        ),
                        standardStartSelect(),
                    ),
                ),
            ),
        ),
        touchProfile(
            familyId = "gba",
            displayName = "Game Boy Advance",
            platformSlugs = setOf("gba"),
            orientationPolicy = PlayerOrientationPolicy.LANDSCAPE_ONLY,
            presets = listOf(
                TouchLayoutPreset(
                    presetId = "standard",
                    displayName = "Standard",
                    elements = listOf(
                        standardDpad(),
                        faceTwo(
                            primaryLabel = "A",
                            secondaryLabel = "B",
                            primaryKey = KeyEvent.KEYCODE_BUTTON_A,
                            secondaryKey = KeyEvent.KEYCODE_BUTTON_B,
                        ).copy(centerY = 0.70f),
                        standardShoulders(labels = "L" to "R"),
                        standardStartSelect(),
                    ),
                ),
            ),
        ),
        touchProfile(
            familyId = "sega16",
            displayName = "Sega 8/16-bit",
            platformSlugs = setOf("genesis", "sms", "gamegear", "segacd"),
            presets = listOf(
                TouchLayoutPreset(
                    presetId = "standard",
                    displayName = "Standard",
                    elements = listOf(
                        standardDpad(),
                        faceFour(
                            left = "B" to KeyEvent.KEYCODE_BUTTON_Y,
                            bottom = "A" to KeyEvent.KEYCODE_BUTTON_B,
                            right = "C" to KeyEvent.KEYCODE_BUTTON_A,
                            top = "X" to KeyEvent.KEYCODE_BUTTON_X,
                        ).copy(centerY = 0.69f),
                        standardStartSelect(selectLabel = "Mode", startLabel = "Start"),
                    ),
                ),
            ),
        ),
        controllerFirstProfile(
            familyId = "sega32x",
            displayName = "Sega 32X",
            platformSlugs = setOf("32x", "sega32x", "sega-32x"),
        ),
        touchProfile(
            familyId = "psx",
            displayName = "PlayStation",
            platformSlugs = setOf("psx", "ps1", "playstation"),
            orientationPolicy = PlayerOrientationPolicy.LANDSCAPE_ONLY,
            presets = listOf(
                TouchLayoutPreset(
                    presetId = "digital",
                    displayName = "Digital",
                    elements = listOf(
                        standardDpad(),
                        faceFour(
                            left = "Square" to KeyEvent.KEYCODE_BUTTON_Y,
                            bottom = "Cross" to KeyEvent.KEYCODE_BUTTON_B,
                            right = "Circle" to KeyEvent.KEYCODE_BUTTON_A,
                            top = "Triangle" to KeyEvent.KEYCODE_BUTTON_X,
                        ),
                        standardShoulders(
                            id = "shoulders_top",
                            labels = "L1" to "R1",
                            keyCodes = KeyEvent.KEYCODE_BUTTON_L1 to KeyEvent.KEYCODE_BUTTON_R1,
                        ),
                        standardShoulders(
                            id = "shoulders_bottom",
                            labels = "L2" to "R2",
                            keyCodes = KeyEvent.KEYCODE_BUTTON_L2 to KeyEvent.KEYCODE_BUTTON_R2,
                        ).copy(centerY = 0.25f, baseScale = 0.84f),
                        standardStartSelect(),
                    ),
                ),
            ),
        ),
        touchProfile(
            familyId = "atari",
            displayName = "Atari / Handheld Classics",
            platformSlugs = setOf("atari2600", "atari7800", "lynx"),
            presets = listOf(
                TouchLayoutPreset(
                    presetId = "standard",
                    displayName = "Standard",
                    elements = listOf(
                        standardDpad(),
                        faceTwo(
                            primaryLabel = "1",
                            secondaryLabel = "2",
                            primaryKey = KeyEvent.KEYCODE_BUTTON_A,
                            secondaryKey = KeyEvent.KEYCODE_BUTTON_B,
                        ),
                        standardStartSelect(selectLabel = "Option", startLabel = "Start"),
                    ),
                ),
            ),
        ),
        touchProfile(
            familyId = "tg16",
            displayName = "TurboGrafx / Neo Geo Pocket / WonderSwan",
            platformSlugs = setOf(
                "tg16",
                "neo-geo-pocket",
                "neo-geo-pocket-color",
                "wonderswan",
                "wonderswan-color",
            ),
            presets = listOf(
                TouchLayoutPreset(
                    presetId = "standard",
                    displayName = "Standard",
                    elements = listOf(
                        standardDpad(),
                        faceTwo(
                            primaryLabel = "II",
                            secondaryLabel = "I",
                            primaryKey = KeyEvent.KEYCODE_BUTTON_A,
                            secondaryKey = KeyEvent.KEYCODE_BUTTON_B,
                        ),
                        standardStartSelect(),
                    ),
                ),
            ),
        ),
        touchProfile(
            familyId = "arcade",
            displayName = "Arcade",
            platformSlugs = setOf("arcade"),
            presets = listOf(
                TouchLayoutPreset(
                    presetId = "standard",
                    displayName = "Standard",
                    elements = listOf(
                        standardDpad(),
                        faceFour(
                            left = "1" to KeyEvent.KEYCODE_BUTTON_Y,
                            bottom = "2" to KeyEvent.KEYCODE_BUTTON_B,
                            right = "3" to KeyEvent.KEYCODE_BUTTON_A,
                            top = "4" to KeyEvent.KEYCODE_BUTTON_X,
                        ),
                        standardStartSelect(selectLabel = "Coin", startLabel = "Start"),
                    ),
                ),
            ),
        ),
        controllerFirstProfile(
            familyId = "n64",
            displayName = "Nintendo 64",
            platformSlugs = setOf("n64"),
        ),
        controllerFirstProfile(
            familyId = "psp",
            displayName = "PlayStation Portable",
            platformSlugs = setOf("psp"),
        ),
        touchProfile(
            familyId = "nds",
            displayName = "Nintendo DS / DSi-enhanced",
            platformSlugs = setOf("nds", "dsi", "nintendo-dsi"),
            orientationPolicy = PlayerOrientationPolicy.PORTRAIT_ONLY,
            presets = listOf(
                TouchLayoutPreset(
                    presetId = "portrait-handheld",
                    displayName = "Portrait handheld",
                    elements = listOf(
                        standardDpad().copy(centerX = 0.19f, centerY = 0.80f),
                        faceFour(
                            left = "Y" to KeyEvent.KEYCODE_BUTTON_Y,
                            bottom = "B" to KeyEvent.KEYCODE_BUTTON_B,
                            right = "A" to KeyEvent.KEYCODE_BUTTON_A,
                            top = "X" to KeyEvent.KEYCODE_BUTTON_X,
                        ).copy(centerX = 0.81f, centerY = 0.79f, baseScale = 0.94f),
                        standardShoulders(labels = "L" to "R").copy(centerY = 0.11f, baseScale = 0.86f),
                        standardStartSelect().copy(centerY = 0.72f, baseScale = 0.82f),
                    ),
                ),
            ),
        ),
        controllerFirstProfile(
            familyId = "dreamcast",
            displayName = "Dreamcast / NAOMI",
            platformSlugs = setOf("dreamcast", "naomi"),
        ),
        controllerFirstProfile(
            familyId = "3do",
            displayName = "3DO",
            platformSlugs = setOf("3do"),
        ),
        controllerFirstProfile(
            familyId = "virtualboy",
            displayName = "Virtual Boy",
            platformSlugs = setOf("virtualboy", "virtual-boy"),
        ),
        controllerFirstProfile(
            familyId = "dolphin",
            displayName = "GameCube / Wii",
            platformSlugs = setOf("gamecube", "ngc", "wii"),
        ),
        controllerFirstProfile(
            familyId = "3ds",
            displayName = "Nintendo 3DS",
            platformSlugs = setOf("3ds", "new-nintendo-3ds"),
        ),
        controllerFirstProfile(
            familyId = "dos",
            displayName = "DOS",
            platformSlugs = setOf("dos"),
            message = "DOS stays controller-first until a validated pointer and keyboard helper overlay exists.",
        ),
    )

    override fun resolve(platformSlug: String): PlatformControlProfile {
        return profiles.firstOrNull { profile ->
            platformSlug.lowercase() in profile.platformSlugs
        } ?: unsupportedProfile(
            familyId = platformSlug.lowercase(),
            displayName = platformSlug.replaceFirstChar { if (it.isLowerCase()) it.titlecase() else it.toString() },
            platformSlugs = setOf(platformSlug.lowercase()),
            message = "This platform is not enabled in the embedded player yet.",
        )
    }

    override fun supportedProfiles(): List<PlatformControlProfile> = profiles

    private fun touchProfile(
        familyId: String,
        displayName: String,
        platformSlugs: Set<String>,
        orientationPolicy: PlayerOrientationPolicy = PlayerOrientationPolicy.AUTO,
        presets: List<TouchLayoutPreset>,
    ): PlatformControlProfile {
        return PlatformControlProfile(
            familyId = familyId,
            displayName = displayName,
            platformSlugs = platformSlugs,
            supportTier = EmbeddedSupportTier.TOUCH_SUPPORTED,
            touchSupportMode = TouchSupportMode.FULL,
            playerOrientationPolicy = orientationPolicy,
            preferredViewportAspectRatio = when (familyId) {
                "nes", "snes", "sega16", "psx", "atari", "tg16", "arcade" -> 4f / 3f
                "gb" -> 10f / 9f
                "gba" -> 3f / 2f
                "nds" -> 2f / 3f
                else -> null
            },
            defaultPresetId = presets.firstOrNull()?.presetId,
            presets = presets,
        )
    }

    private fun controllerFirstProfile(
        familyId: String,
        displayName: String,
        platformSlugs: Set<String>,
        message: String = "Connect a controller for this platform. Touch support comes later after a platform-specific layout is validated.",
    ): PlatformControlProfile {
        return PlatformControlProfile(
            familyId = familyId,
            displayName = displayName,
            platformSlugs = platformSlugs,
            supportTier = EmbeddedSupportTier.CONTROLLER_SUPPORTED,
            touchSupportMode = TouchSupportMode.CONTROLLER_FIRST,
            playerOrientationPolicy = PlayerOrientationPolicy.AUTO,
            controllerFallbackMessage = message,
        )
    }

    private fun unsupportedProfile(
        familyId: String,
        displayName: String,
        platformSlugs: Set<String>,
        message: String,
    ): PlatformControlProfile {
        return PlatformControlProfile(
            familyId = familyId,
            displayName = displayName,
            platformSlugs = platformSlugs,
            supportTier = EmbeddedSupportTier.UNSUPPORTED,
            touchSupportMode = TouchSupportMode.CONTROLLER_FIRST,
            playerOrientationPolicy = PlayerOrientationPolicy.AUTO,
            controllerFallbackMessage = message,
        )
    }
}
